src/test/hdmint_tests.cpp

```diff
@@ -150,4 +155,93 @@ BOOST_AUTO_TEST_CASE(deterministic)
+*/
+BOOST_AUTO_TEST_CASE(wallet_count_tests)
+{
+    vector<string> denominationsForTx;
+    vector<uint256> vtxid;
+    string thirdPartyAddress;
+    int previousHeight;
+    CBlock b;
+    CWalletTx wtx;
+
+    const int TOTAL_MINTS = 5;
+    const int INITIAL_MINTS = TOTAL_MINTS-1;
+
+    sigma::CSigmaState *sigmaState = sigma::CSigmaState::GetState();
+
+    // Create 400-200+1 = 201 new empty blocks. // consensus.nMintV3SigmaStartBlock = 400
+    CreateAndProcessEmptyBlocks(201, scriptPubKey);
+
+    pwalletMain->SetBroadcastTransactions(true);
+
+    for(int i=0; i<=INITIAL_MINTS;i++){
+        CAmount nAmount = AmountFromValue("1");
+
+        std::vector<sigma::CoinDenomination> denominations;
+        sigma::GetAllDenoms(denominations);
+
+        CAmount smallestDenom;
+        DenominationToInteger(denominations.back(), smallestDenom);
+
+        if (nAmount % smallestDenom != 0) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Amount to mint is invalid.\n");
+        }
+
+        std::vector<sigma::CoinDenomination> mints;
+
+        BOOST_CHECK(CWallet::SelectMintCoinsForAmount(nAmount, denominations, mints) == nAmount);
+
+        std::vector<sigma::PrivateCoin> privCoins;
+
+        const auto& sigmaParams = sigma::Params::get_default();
+        std::transform(mints.begin(), mints.end(), std::back_inserter(privCoins),
+            [sigmaParams](const sigma::CoinDenomination& denom) -> sigma::PrivateCoin {
+                return sigma::PrivateCoin(sigmaParams, denom);
+            });
+        vector<CHDMint> vDMints;
+        auto vecSend = CWallet::CreateSigmaMintRecipients(privCoins, vDMints);
+
+        CWalletTx wtx;
+        std::string strError = pwalletMain->MintAndStoreSigma(vecSend, privCoins, vDMints, wtx);
+
+        BOOST_CHECK(strError == "");
+    }
+
+    zwalletMain->SetCount(0); // reset count
+
+    // create another mint
+    CAmount nAmount = AmountFromValue("1");
+
+    std::vector<sigma::CoinDenomination> denominations;
+    sigma::GetAllDenoms(denominations);
+
+    CAmount smallestDenom;
+    DenominationToInteger(denominations.back(), smallestDenom);
+
+    if (nAmount % smallestDenom != 0) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Amount to mint is invalid.\n");
+    }
+
+    std::vector<sigma::CoinDenomination> mints;
+    BOOST_CHECK(CWallet::SelectMintCoinsForAmount(nAmount, denominations, mints) == nAmount);
+
+    std::vector<sigma::PrivateCoin> privCoins;
+
+    const auto& sigmaParams = sigma::Params::get_default();
+    std::transform(mints.begin(), mints.end(), std::back_inserter(privCoins),
+    [sigmaParams](const sigma::CoinDenomination& denom) -> sigma::PrivateCoin {
+        return sigma::PrivateCoin(sigmaParams, denom);
+    });
+    vector<CHDMint> vDMints;
+    auto vecSend = CWallet::CreateSigmaMintRecipients(privCoins, vDMints);
+
+    BOOST_CHECK(vDMints[0].GetCount() == TOTAL_MINTS);
+    BOOST_CHECK(zwalletMain->GetCount() == (TOTAL_MINTS+1));
+
+}
+
```

Comment:
I think we need a test where you are given a blockchain already having mints generated with current seed and future values of the counter. Then we need to ensure the code is picking that up correctly and generates next mint with no collision with the ones already in the blockchain. This is the current issue with mint creation after all.

Reply:
Agree. Working on it